File: hostel_management_system_backend/Controllers/AuthController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using hostel_management_system_backend.Exceptions;

[ApiController]
[Route("api/[controller]")]
public sealed class AuthController : ControllerBase
{
    private readonly IAuthService _authService;
    private readonly IConfiguration _configuration;

    public AuthController(IAuthService authService, IConfiguration configuration)
    {
        _authService = authService;
        _configuration = configuration;
    }

    [AllowAnonymous]
    [HttpPost("login")]
    public async Task<ActionResult<AuthTokensResponseDto>> Login([FromBody] LoginRequestDto dto, CancellationToken cancellationToken)
    {
        var result = await _authService.LoginAsync(dto, cancellationToken)
            ?? throw new UnauthorizedException("Invalid email or password.");

        WriteRefreshCookie(result.RefreshToken, result.RefreshTokenExpiresAt, result.Response.Role == UserRole.Admin);
        return Ok(result.Response);
    }

    [AllowAnonymous]
    [HttpPost("register")]
    public async Task<ActionResult<AuthTokensResponseDto>> Register([FromBody] UserRegisterDto dto, CancellationToken cancellationToken)
    {
        var result = await _authService.RegisterAsync(dto, cancellationToken);

        WriteRefreshCookie(result.RefreshToken, result.RefreshTokenExpiresAt);
        return CreatedAtAction(nameof(Register), result.Response);
    }

    [AllowAnonymous]
    [HttpPost("refresh")]
    public async Task<ActionResult<AuthTokensResponseDto>> Refresh(CancellationToken cancellationToken)
    {
        var refreshToken = ReadRefreshCookie();
        if (string.IsNullOrWhiteSpace(refreshToken))
        {
            return Unauthorized("Refresh token is missing.");
        }

        AuthTokenIssueResult result;
        try
        {
            result = await _authService.RefreshAsync(refreshToken, cancellationToken)
                ?? throw new UnauthorizedException("Session expired. Please log in again.");
        }
        catch (UnauthorizedException)
        {
            DeleteRefreshCookie();
            throw;
        }

        WriteRefreshCookie(result.RefreshToken, result.RefreshTokenExpiresAt, result.Response.Role == UserRole.Admin);
        return Ok(result.Response);
    }

    [HttpPost("logout")]
    public async Task<IActionResult> Logout(CancellationToken cancellationToken)
    {
        var refreshToken = ReadRefreshCookie();
        if (!string.IsNullOrWhiteSpace(refreshToken))
        {
            await _authService.LogoutAsync(refreshToken, cancellationToken);
        }

        DeleteRefreshCookie();
        return NoContent();
    }

    private string GetRefreshCookieName()
    {
        return _configuration["AuthSettings:RefreshCookieName"] ?? "refreshToken";
    }

    private string? ReadRefreshCookie()
    {
        Request.Cookies.TryGetValue(GetRefreshCookieName(), out var refreshToken);
        return refreshToken;
    }

    private void WriteRefreshCookie(string token, DateTime expiresAt, bool sessionOnly = false)
    {
        var options = new CookieOptions
        {
            HttpOnly = true,
            Secure = Request.IsHttps,
            SameSite = SameSiteMode.Strict
        };

        if (!sessionOnly)
        {
            options.Expires = new DateTimeOffset(expiresAt);
        }

        Response.Cookies.Append(GetRefreshCookieName(), token, options);
    }

    private void DeleteRefreshCookie()
    {
        Response.Cookies.Delete(GetRefreshCookieName());
    }
}

File: src/auth/auth.controller.ts
import { Body, Controller, HttpCode, HttpStatus, Post, Req, Res } from '@nestjs/common';
import { Request, Response } from 'express';

import { UserRole } from '../common/enums/app.enums';
import { AppConfigService } from '../config/app-config.service';
import { UserRegisterDto } from '../users/dto/users.dto';
import { AuthTokensResponseDto, LoginRequestDto } from './dto/auth.dto';
import { AuthService } from './auth.service';

@Controller('auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly configService: AppConfigService,
  ) {}

  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(
    @Body() dto: LoginRequestDto,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthTokensResponseDto> {
    const result = await this.authService.login(dto);
    this.writeRefreshCookie(response, result.refreshToken, result.refreshTokenExpiresAt, result.response.role === UserRole.Admin);
    return result.response;
  }

  @Post('register')
  async register(
    @Body() dto: UserRegisterDto,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthTokensResponseDto> {
    const result = await this.authService.register(dto);
    this.writeRefreshCookie(response, result.refreshToken, result.refreshTokenExpiresAt, false);
    response.status(HttpStatus.CREATED);
    return result.response;
  }

  @Post('refresh')
  @HttpCode(HttpStatus.OK)
  async refresh(
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ): Promise<AuthTokensResponseDto> {
    const refreshToken = request.cookies?.[this.configService.refreshCookieName];
    const result = await this.authService.refresh(refreshToken);
    this.writeRefreshCookie(response, result.refreshToken, result.refreshTokenExpiresAt, result.response.role === UserRole.Admin);
    return result.response;
  }

  @Post('logout')
  @HttpCode(HttpStatus.NO_CONTENT)
  async logout(
    @Req() request: Request,
    @Res({ passthrough: true }) response: Response,
  ) {
    const refreshToken = request.cookies?.[this.configService.refreshCookieName];
    await this.authService.logout(refreshToken);
    response.clearCookie(this.configService.refreshCookieName);
    response.status(HttpStatus.NO_CONTENT);
  }

  private writeRefreshCookie(response: Response, token: string, expiresAt: Date, sessionOnly: boolean) {
    response.cookie(this.configService.refreshCookieName, token, {
      httpOnly: true,
      sameSite: 'strict',
      secure: this.configService.nodeEnv === 'production',
      ...(sessionOnly ? {} : { expires: expiresAt }),
    });
  }
}
